docs/0002-events-architecture.md

```diff
@@ -0,0 +1,84 @@
+Events will be emitted when users interact with the component (e.g., clicking a button or selecting an option from a dropdown). 
+These events can then be listened to by any part of the application to trigger GA data layer pushes or other actions, without requiring changes to the Visual Editor component itself.
+
+## Decision
+
+We will implement an Event Emitter system in the Visual Editor. The component will emit events for user interactions such as button clicks and dropdown selections. 
+These events will carry necessary data attributes, and other parts of the application can listen to these events to perform actions like GA tracking.
+
+The following steps outline the implementation:
+
+1. Create an `EventEmitter` class to manage event listeners and emit events.
+2. Integrate the `EventEmitter` into the `VisualEditor` component.
+3. Emit events from the `VisualEditor` component when user interactions occur (e.g., `clickButton`, `selectOption`).
+4. Other parts of the application can register event listeners to handle these events and push data to the GA data layer as needed.
+
+The proposed class diagram is as follows:
+
+```mermaid
+classDiagram
+    class EventEmitter {
+        +events: Object
+        +on(event: String, listener: Function): void
+        +off(event: String, listenerToRemove: Function): void
+        +emit(event: String, data: Object): void
+    }
+
+    class VisualEditor {
+        +clickButton(): void
+        +selectOption(): void
+    }
+    
+    EventEmitter <|-- VisualEditor
+    
+    class User {
+        +visualEditor: VisualEditor
+        +onClickButton(data: Object): void
+        +onSelectOption(data: Object): void
+    }
+    
+    VisualEditor --> "1..*" EventEmitter : emits
+    User --> "1" VisualEditor : listens to
+    
+    EventEmitter : +events
+    EventEmitter : +on(event, listener)
+    EventEmitter : +off(event, listenerToRemove)
+    EventEmitter : +emit(event, data)
+    
+    VisualEditor : +clickButton()
+    VisualEditor : +selectOption()
+    
+    User : +onClickButton(data)
+    User : +onSelectOption(data)
+```
+
+## Consequences
+
+### Benefits:
+- **Decoupling**: GA tracking logic is decoupled from the Visual Editor component, allowing for easier maintenance and updates.
+- **Flexibility**: Any part of the application can listen to events and perform actions, making it easier to extend functionality without modifying the Visual Editor component.
+- **Simplicity**: Reduces the complexity of the Visual Editor component by removing hardcoded GA tracking logic.
+
+### Risks:
+- **Event Handling Overhead**: Introducing an event system adds complexity in terms of event management and debugging.
+- **Performance Impact**: Emitting and listening to events could introduce a slight performance overhead, especially if not managed properly.
+- **Error Handling**: Ensuring that event listeners handle errors gracefully to avoid breaking the application flow.
```

Comment:
Do we want to do too much for error handling? If a consumer of the visual editor library passes a function which results in an error, is it the responsibility of the editor to try to handle the error? I would say "no" at first glance but I might be missing something.

Reply:
Yeah I guess its their responsibility, just make sure that calling a callback that results an error doest affect our own code.